### PR TITLE
Feature/prometheus

### PR DESCRIPTION
**Pull request by vsaveliev**

### CODE REVIEW

File: inventory/group_vars/all.yml [hunk 1]
```diff
@@ -301,7 +305,7 @@ k8s_services_cert_key: |
 # https://developers.google.com/identity/protocols/OAuth2ServiceAccount#creatinganaccount
 gce_service_account_email: '...-compute@developer.gserviceaccount.com'
 
-# The full path of your unique service account credentials file.
```
Comment (takama) [inline inventory/group_vars/all.yml:392]: Things like that are much convenient in separate PR

File: roles/prometheus/defaults/main.yml [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+k8s_master_hosts: {}
+
+# Prometheus domains (for Prometheus, AlertManager, PushGateway)
+k8s_prometheus_name: ''
+k8s_prometheus_alertmanager_name: ''
+k8s_prometheus_pushgateway_name: ''
+
+# Prometheus images
+k8s_prometheus_image: prom/prometheus
+k8s_prometheus_image_tag: v1.5.1
```
Comment (takama) [inline roles/prometheus/defaults/main.yml]: Why we use 8 month old image tag?

File: roles/prometheus/defaults/main.yml [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+k8s_prometheus_alertmanager_image: prom/alertmanager
+k8s_prometheus_alertmanager_image_tag: v0.5.1
```
Comment (takama) [inline roles/prometheus/defaults/main.yml]: 0.9.1

File: roles/prometheus/defaults/main.yml [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+k8s_prometheus_ksm_image: gcr.io/google_containers/kube-state-metrics
+k8s_prometheus_ksm_image_tag: v0.4.1
```
Comment (takama) [inline roles/prometheus/defaults/main.yml]: v1.0.1

File: roles/prometheus/defaults/main.yml [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+k8s_prometheus_nodeexport_image: prom/node-exporter
+k8s_prometheus_nodeexport_image_tag: v0.13.0
```
Comment (takama) [inline roles/prometheus/defaults/main.yml]: v0.14.0

File: roles/prometheus/handlers/main.yml [hunk 6]
```diff
@@ -0,0 +1,2 @@
+---
+# handlers file for prometheus
```
Comment (takama) [inline roles/prometheus/handlers/main.yml]: CR

File: roles/prometheus/meta/main.yml [hunk 7]
```diff
@@ -0,0 +1,57 @@
+galaxy_info:
+  author: your name
+  description: your description
+  company: your company (optional)
```
Comment (takama) [inline roles/prometheus/meta/main.yml]: Please fill that

File: roles/prometheus/templates/alert-manager.yaml [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+      labels:
+        app: prometheus
+        component: "alertmanager"
+    spec:
+      serviceAccountName: prometheus-alertmanager
+      containers:
+        - name: prometheus-alertmanager
+          image: "{{ k8s_prometheus_alertmanager_image }}:{{ k8s_prometheus_alertmanager_image_tag }}"
+          imagePullPolicy: "IfNotPresent"
+          env:
```
Comment (takama) [inline roles/prometheus/templates/alert-manager.yaml]: Why we need empty env?

File: roles/prometheus/templates/alert-manager.yaml [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+            - name: storage-volume
+              mountPath: "/data"
+              subPath: ""
+        - name: prometheus-alertmanager-configmap-reload
+          image: "{{ k8s_prometheus_configreload_image }}:{{ k8s_prometheus_configreload_image_tag }}"
+          imagePullPolicy: "IfNotPresent"
+          args:
+            - --volume-dir=/etc/config
+            - --webhook-url=http://localhost:9093/-/reload
+          resources:
```
Comment (takama) [inline roles/prometheus/templates/alert-manager.yaml]: Why we need empty resources?

File: roles/prometheus/templates/kube-state-metrics.yaml [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+  labels:
+    app: prometheus
+    release: prometheus-{{ k8s_prometheus_image_tag }}
+    component: "kube-state-metrics"
+  namespace: {{ k8s_prometheus_namespace }}
+  name: prometheus-kube-state-metrics
+subjects:
+  - kind: ServiceAccount
+    name: prometheus-kube-state-metrics
+    namespace: default
```
Comment (takama) [inline roles/prometheus/templates/kube-state-metrics.yaml]: incorrect namespace, should be {{ k8s_prometheus_namespace }}

File: roles/prometheus/templates/kube-state-metrics.yaml [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+---
+# Source: prometheus/templates/kube-state-metrics-clusterrolebinding.yaml
+apiVersion: rbac.authorization.k8s.io/v1beta1
+kind: ClusterRoleBinding
+metadata:
+  labels:
+    app: prometheus
+    release: prometheus-{{ k8s_prometheus_image_tag }}
+    component: "kube-state-metrics"
+  namespace: {{ k8s_prometheus_namespace }}
```
Comment (takama) [inline roles/prometheus/templates/kube-state-metrics.yaml]: ClusterRoleBinding can not have any namespace

File: roles/prometheus/templates/alert-manager.yaml [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+---
+# Source: prometheus/templates/alertmanager-clusterrolebinding.yaml
+apiVersion: rbac.authorization.k8s.io/v1beta1
+kind: ClusterRoleBinding
+metadata:
+  labels:
+    app: prometheus
+    release: prometheus-{{ k8s_prometheus_image_tag }}
+    component: "alertmanager"
+  namespace: {{ k8s_prometheus_namespace }}
```
Comment (takama) [inline roles/prometheus/templates/alert-manager.yaml]: ClusterRoleBinding can not have any namespace

File: roles/prometheus/templates/alert-manager.yaml [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
+  name: prometheus-alertmanager
+subjects:
+  - kind: ServiceAccount
+    name: prometheus-alertmanager
+    namespace: default
```
Comment (takama) [inline roles/prometheus/templates/alert-manager.yaml]: incorrect namespace, should be {{ k8s_prometheus_namespace }}

File: roles/prometheus/templates/kube-state-metrics.yaml [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+    spec:
+      serviceAccountName: prometheus-kube-state-metrics
+      containers:
+        - name: prometheus-kube-state-metrics
+          image: "{{ k8s_prometheus_ksm_image }}:{{ k8s_prometheus_ksm_image_tag }}"
+          imagePullPolicy: "IfNotPresent"
+          ports:
+            - name: metrics
+              containerPort: 8080
+          resources:
```
Comment (takama) [inline roles/prometheus/templates/kube-state-metrics.yaml:128]: empty resources

File: roles/prometheus/templates/node-exporter.yaml [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+metadata:
+  labels:
+    app: prometheus
+    release: prometheus-{{ k8s_prometheus_image_tag }}
+    component: "node-exporter"
+  name: prometheus-node-exporter
+subjects:
+  - kind: ServiceAccount
+    name: prometheus-node-exporter
+    namespace: default
```
Comment (takama) [inline roles/prometheus/templates/node-exporter.yaml]: incorrect namespace

File: roles/prometheus/templates/node-exporter.yaml [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+          image: "{{ k8s_prometheus_nodeexport_image }}:{{ k8s_prometheus_nodeexport_image_tag }}"
+          imagePullPolicy: "IfNotPresent"
+          args:
+            - --collector.procfs=/host/proc
+            - --collector.sysfs=/host/sys
+          ports:
+            - name: metrics
+              containerPort: 9100
+              hostPort: 9100
+          resources:
```
Comment (takama) [inline roles/prometheus/templates/node-exporter.yaml]: empty resources

File: roles/prometheus/templates/push-gateway.yaml [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+    metadata:
+      labels:
+        app: prometheus
+        component: "pushgateway"
+    spec:
+      containers:
+        - name: prometheus-pushgateway
+          image: "{{ k8s_prometheus_pushgateway_image }}:{{ k8s_prometheus_pushgateway_image_tag }}"
+          imagePullPolicy: "IfNotPresent"
+          args:
```
Comment (takama) [inline roles/prometheus/templates/push-gateway.yaml]: empty args

File: roles/prometheus/templates/push-gateway.yaml [hunk 18]
```diff
@@ -0,0 +1,85 @@
...
+          ports:
+            - containerPort: 9091
+          readinessProbe:
+            httpGet:
+              path: /#/status
+              port: 9091
+            initialDelaySeconds: 10
+            timeoutSeconds: 10
+          resources:
```
Comment (takama) [inline roles/prometheus/templates/push-gateway.yaml:54]: empty resources

File: roles/prometheus/templates/server.yaml [hunk 19]
```diff
@@ -0,0 +1,210 @@
...
+---
+# Source: prometheus/templates/server-clusterrolebinding.yaml
+apiVersion: rbac.authorization.k8s.io/v1beta1
+kind: ClusterRoleBinding
+metadata:
+  labels:
+    app: prometheus
+    release: prometheus-{{ k8s_prometheus_image_tag }}
+    component: "server"
+  namespace: {{ k8s_prometheus_namespace }}
```
Comment (takama) [inline roles/prometheus/templates/server.yaml]: ClusterRoleBinding can not have any namespace

File: roles/prometheus/templates/server.yaml [hunk 20]
```diff
@@ -0,0 +1,210 @@
...
+    spec:
+      serviceAccountName: prometheus-server
+      containers:
+        - name: prometheus-server-configmap-reload
+          image: "{{ k8s_prometheus_configreload_image }}:{{ k8s_prometheus_configreload_image_tag }}"
+          imagePullPolicy: "IfNotPresent"
+          args:
+            - --volume-dir=/etc/config
+            - --webhook-url=http://localhost:9090/-/reload
+          resources:
```
Comment (takama) [inline roles/prometheus/templates/server.yaml]: empty resources

File: roles/prometheus/templates/server.yaml [hunk 21]
```diff
@@ -0,0 +1,210 @@
...
+            - --web.console.templates=/etc/prometheus/consoles
+          ports:
+            - containerPort: 9090
+          readinessProbe:
+            httpGet:
+              path: /status
+              port: 9090
+            initialDelaySeconds: 30
+            timeoutSeconds: 30
+          resources:
```
Comment (takama) [inline roles/prometheus/templates/server.yaml]: empty resources

File: roles/prometheus/vars/main.yml [hunk 22]
```diff
@@ -0,0 +1,2 @@
+---
+# vars file for prometheus
```
Comment (takama) [inline roles/prometheus/vars/main.yml]: CR

**Review by takama — CHANGES_REQUESTED** [review]

File: inventory/group_vars/all.yml (hunk 1 above)
Comment (vsaveliev) [reply inventory/group_vars/all.yml:392]: ok, next time I will create separate PR for it.

File: roles/prometheus/handlers/main.yml (hunk 6 above)
Comment (vsaveliev) [reply roles/prometheus/handlers/main.yml]: fixed

File: roles/prometheus/templates/alert-manager.yaml (hunk 12 above)
Comment (vsaveliev) [reply roles/prometheus/templates/alert-manager.yaml]: fixed

File: roles/prometheus/templates/kube-state-metrics.yaml (hunk 11 above)
Comment (vsaveliev) [reply roles/prometheus/templates/kube-state-metrics.yaml]: fixed

File: roles/prometheus/templates/server.yaml (hunk 19 above)
Comment (vsaveliev) [reply roles/prometheus/templates/server.yaml]: fixed

File: roles/prometheus/vars/main.yml (hunk 22 above)
Comment (vsaveliev) [reply roles/prometheus/vars/main.yml]: fixed

**Comment by vsaveliev** [comment]
image versions were extracted from stable charts. I forgot to update and check it. I have updated it.

File: roles/prometheus/templates/alert-manager.yaml [hunk 23]
```diff
@@ -0,0 +1,172 @@
...
+      - name: slack-receiver
+        slack_configs:
+          - channel: '{{ prometheus_slack_channel }}'
+            send_resolved: true
+{% endif %}
+
+    route:
+      group_wait: 10s
+      group_interval: 5m
+      receiver: slack-receiver
```
Comment (IKSIN) [inline roles/prometheus/templates/alert-manager.yaml]: If slack-receiver is not defined, we will still try to use it.

**Review by IKSIN — APPROVED** [review]
LGFM

**Comment by vsaveliev** [comment]
Added grafana w/o imported dashboards.

**Review by takama — APPROVED** [review]
LGTM

**Review by STASiAN — APPROVED** [review]

File: roles/prometheus/templates/alert-manager.yaml [hunk 24]
```diff
@@ -0,0 +1,205 @@
...
+metadata:
+  labels:
+    app: prometheus
+    component: "alertmanager"
+  namespace: {{ k8s_prometheus_namespace }}
+  name: prometheus-alertmanager
+
+---
+
+apiVersion: rbac.authorization.k8s.io/v1beta1
```
Comment (takama) [inline roles/prometheus/templates/alert-manager.yaml]: It will be `v1`

File: roles/prometheus/templates/blackbox-exporter.yaml [hunk 25]
```diff
@@ -0,0 +1,166 @@
...
+metadata:
+  labels:
+    app: prometheus
+    component: "blackbox-exporter"
+  namespace: {{ k8s_prometheus_namespace }}
+  name: prometheus-blackbox-exporter
+
+---
+
+apiVersion: rbac.authorization.k8s.io/v1beta1
```
Comment (takama) [inline roles/prometheus/templates/blackbox-exporter.yaml]: It will be `v1`

File: roles/prometheus/templates/config.yaml [hunk 26]
```diff
@@ -0,0 +1,60 @@
...
+metadata:
+  name: registry-pull-secret
+  namespace: {{ k8s_prometheus_namespace }}
+type: kubernetes.io/dockerconfigjson
+data:
+  .dockerconfigjson: {{ k8s_docker_registry_auth_token }}
+
+---
+
+apiVersion: rbac.authorization.k8s.io/v1beta1
```
Comment (takama) [inline roles/prometheus/templates/config.yaml]: It will be `v1`

File: roles/prometheus/templates/config.yaml [hunk 27]
```diff
@@ -0,0 +1,60 @@
...
+- apiGroups:
+  - '*'
+  resources:
+  - '*'
+  verbs:
+  - '*'
+
+---
+
+apiVersion: rbac.authorization.k8s.io/v1beta1
```
Comment (takama) [inline roles/prometheus/templates/config.yaml]: It will be `v1`

File: roles/prometheus/templates/kube-state-metrics.yaml [hunk 28]
```diff
@@ -0,0 +1,134 @@
...
+metadata:
+  labels:
+    app: prometheus
+    component: "kube-state-metrics"
+  namespace: {{ k8s_prometheus_namespace }}
+  name: prometheus-kube-state-metrics
+
+---
+
+apiVersion: rbac.authorization.k8s.io/v1beta1
```
Comment (takama) [inline roles/prometheus/templates/kube-state-metrics.yaml]: It will be `v1`

File: roles/prometheus/templates/kube-state-metrics.yaml [hunk 29]
```diff
@@ -0,0 +1,134 @@
...
+    resources:
+      - cronjobs
+      - jobs
+    verbs:
+      - list
+      - watch
+
+---
+
+apiVersion: rbac.authorization.k8s.io/v1beta1
```
Comment (takama) [inline roles/prometheus/templates/kube-state-metrics.yaml]: It will be `v1`

File: roles/prometheus/templates/node-exporter.yaml [hunk 30]
```diff
@@ -0,0 +1,114 @@
...
+metadata:
+  labels:
+    app: prometheus
+    component: "node-exporter"
+  namespace: {{ k8s_prometheus_namespace }}
+  name: prometheus-node-exporter
+
+---
+
+apiVersion: rbac.authorization.k8s.io/v1beta1
```
Comment (takama) [inline roles/prometheus/templates/node-exporter.yaml]: It will be `v1`

File: roles/prometheus/templates/server.yaml [hunk 31]
```diff
@@ -0,0 +1,217 @@
...
+metadata:
+  labels:
+    app: prometheus
+    component: "server"
+  namespace: {{ k8s_prometheus_namespace }}
+  name: prometheus-server
+
+---
+
+apiVersion: rbac.authorization.k8s.io/v1beta1
```
Comment (takama) [inline roles/prometheus/templates/server.yaml]: It will be `v1`

File: roles/prometheus/templates/server.yaml [hunk 32]
```diff
@@ -0,0 +1,217 @@
...
+      - list
+      - watch
+  - nonResourceURLs:
+      - "/metrics"
+    verbs:
+      - get
+
+---
+
+apiVersion: rbac.authorization.k8s.io/v1beta1
```
Comment (takama) [inline roles/prometheus/templates/server.yaml]: It will be `v1`

File: playbooks/addons/prometheus.yml [hunk 33]
```diff
@@ -0,0 +1,8 @@
+---
+- hosts: addons
+  remote_user: '{{ k8s_ssh_user }}'
+  become: true
+  roles:
+    - role: prometheus
```
Comment (takama) [inline playbooks/addons/prometheus.yml:6]: Could me install prometheus services scrapes in separate roles?

File: roles/prometheus/defaults/main.yml [hunk 34]
```diff
@@ -0,0 +1,81 @@
+---
+# Kubernetes configs path
+k8s_conf_dir: /etc/kubernetes
+k8s_prometheus_dir: '{{ k8s_conf_dir }}/addons/prometheus'
+
+# Master hosts names
+k8s_master_hosts: {}
+
+# Retention days (how long to store data)
+k8s_prometheus_retention_days: 7
```
Comment (takama) [inline roles/prometheus/defaults/main.yml:10]: We should add this parameter with comments in `group_vars/all.yml` also

File: playbooks/addons/prometheus.yml (hunk 33 above)
Comment (vsaveliev) [reply playbooks/addons/prometheus.yml:6]: I will do it next pull request, ok?

File: roles/prometheus/defaults/main.yml (hunk 34 above)
Comment (vsaveliev) [reply roles/prometheus/defaults/main.yml:10]: ok

File: roles/prometheus/templates/alert-manager.yaml (hunk 24 above)
Comment (vsaveliev) [reply roles/prometheus/templates/alert-manager.yaml]: ok

File: roles/prometheus/templates/config.yaml (hunk 26 above)
Comment (vsaveliev) [reply roles/prometheus/templates/config.yaml]: ok

File: roles/prometheus/templates/kube-state-metrics.yaml (hunk 28 above)
Comment (vsaveliev) [reply roles/prometheus/templates/kube-state-metrics.yaml]: ok

File: roles/prometheus/templates/node-exporter.yaml (hunk 30 above)
Comment (vsaveliev) [reply roles/prometheus/templates/node-exporter.yaml]: ok

File: roles/prometheus/templates/kube-state-metrics.yaml (hunk 29 above)
Comment (vsaveliev) [reply roles/prometheus/templates/kube-state-metrics.yaml]: ok

File: roles/prometheus/templates/server.yaml (hunk 31 above)
Comment (vsaveliev) [reply roles/prometheus/templates/server.yaml]: ok

File: roles/prometheus/templates/server.yaml (hunk 32 above)
Comment (vsaveliev) [reply roles/prometheus/templates/server.yaml]: ok

File: roles/prometheus/templates/config.yaml (hunk 27 above)
Comment (vsaveliev) [reply roles/prometheus/templates/config.yaml]: ok

File: roles/prometheus/templates/blackbox-exporter.yaml (hunk 25 above)
Comment (vsaveliev) [reply roles/prometheus/templates/blackbox-exporter.yaml]: ok

**Pull request merged**